python/src/source.rs

```diff
@@ -0,0 +1,161 @@
+        subsort_column: Option<&str>,
+        grouping_name: Option<&str>,
+        time_unit: Option<&str>,
+        source: Option<&str>,
+    ) -> Result<Self> {
+        let raw_schema = Arc::new(schema.0);
+
+        let rust_table = session.rust_session()?.add_table(
+            &name,
+            raw_schema,
+            time_column,
+            queryable,
+            subsort_column,
+            key_column,
+            grouping_name,
+            time_unit,
+            source,
+        )?;
+
+        let table = Self {
+            rust_table: Arc::new(rust_table),
+            session,
+        };
+        Ok(table)
+    }
+
+    fn expr(&self) -> Expr {
+        let rust_expr = self.rust_table.expr.clone();
+        Expr {
+            rust_expr,
+            session: self.session.clone(),
+        }
+    }
+
+    /// Add PyArrow data to the given table.
+    ///
+    /// TODO: Support other kinds of data:
+    /// - pyarrow RecordBatchReader
+    /// - Parquet file URLs
+    /// - Python generators?
+    /// TODO: Error handling
+    fn add_pyarrow<'py>(&self, data: &'py PyAny, py: Python<'py>) -> Result<&'py PyAny> {
+        let data = RecordBatch::from_pyarrow(data)?;
+
+        let rust_table = self.rust_table.clone();
+        Ok(pyo3_asyncio::tokio::future_into_py(py, async move {
+            let result = rust_table.add_data(data).await;
+            Python::with_gil(|py| {
+                result.unwrap();
+                Ok(py.None())
+            })
+        })?)
+    }
+
+    fn add_parquet<'py>(&mut self, py: Python<'py>, path: String) -> Result<&'py PyAny> {
+        let rust_table = self.rust_table.clone();
+        Ok(pyo3_asyncio::tokio::future_into_py(py, async move {
+            let path = std::path::Path::new(&path);
+            let result = rust_table.add_parquet(path).await;
+            Python::with_gil(|py| {
+                result.unwrap();
+                Ok(py.None())
+            })
+        })?)
+    }
+}
+
+#[pyclass]
+pub(crate) struct ParquetSource {
+    rust_table: Arc<RustTable>,
+    session: Session,
+}
+
+#[pymethods]
+impl ParquetSource {
+    /// Create a new table.
+    #[new]
+    #[pyo3(signature = (session, name, time_column, key_column, schema, queryable, subsort_column, grouping_name, time_unit, source))]
+    #[allow(clippy::too_many_arguments)]
+    fn new(
+        session: Session,
+        name: String,
+        time_column: &str,
+        key_column: &str,
+        subsort_column: Option<&str>,
+        grouping_name: Option<&str>,
+        time_unit: Option<&str>,
+    ) -> Result<Self> {
+        // Look at the parquet files.
+        let raw_schema = Arc::new(schema.0);
+
+        let rust_table = session.rust_session()?.add_table(
+            &name,
+            raw_schema,
+            time_column,
+            false,
+            subsort_column,
+            key_column,
+            grouping_name,
+            time_unit,
+            Some("parquet"),
```

Comment:
Should we think about a type for our `source`? 

What happens if the table adds non-parquet files, at what point would it fail?

Reply:
I really want to, but wasn't yet able to. The problem is that under the hood, everything is crammed into the `TableInfo` in the `DataContext`. To simplify things, I think it would be easiest if we wait until we get more of the partitioned / logical plan compilation / etc. since it should be easier to separate the sources based on the source trait you're adding, etc.